File: 351004/Nepesov/Lab1/untitled/src/main/java/org/example/exception/GlobalExceptionHandler.java
package org.example.exception;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

@ControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(EntityNotFoundException.class)
    public ResponseEntity<ErrorResponse> handleNotFound(EntityNotFoundException ex) {
        // Код 40401 для Not Found
        return new ResponseEntity<>(new ErrorResponse(ex.getMessage(), "40401"), HttpStatus.NOT_FOUND);
    }

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ErrorResponse> handleValidation(MethodArgumentNotValidException ex) {
        // Код 40001 для ошибок валидации (@Size и т.д.)
        String message = "Validation failed";
        if (ex.getBindingResult().getFieldError() != null) {
            message = ex.getBindingResult().getFieldError().getDefaultMessage();
        }
        return new ResponseEntity<>(new ErrorResponse(message, "40001"), HttpStatus.BAD_REQUEST);
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ErrorResponse> handleGeneral(Exception ex) {
        // Код 50000 для всех остальных ошибок
        return new ResponseEntity<>(new ErrorResponse("Internal server error", "50000"), HttpStatus.INTERNAL_SERVER_ERROR);
    }
}

File: 351002/Abrashin/publisher/src/main/kotlin/com/example/distcomp/exception/GlobalExceptionHandler.kt
package com.example.distcomp.exception

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler

import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.web.client.RestClientResponseException

@ControllerAdvice
class GlobalExceptionHandler {


    @ExceptionHandler(NotFoundException::class)
    fun handleNotFound(ex: NotFoundException): ResponseEntity<ErrorResponse> {
        return ResponseEntity(ErrorResponse(ex.message ?: "Not Found", ERR_NOT_FOUND), HttpStatus.NOT_FOUND)
    }

    @ExceptionHandler(ConflictException::class)
    fun handleConflict(ex: ConflictException): ResponseEntity<ErrorResponse> {
        return ResponseEntity(ErrorResponse(ex.message ?: "Conflict", ERR_FORBIDDEN), HttpStatus.FORBIDDEN)
    }


    @ExceptionHandler(BadRequestException::class)
    fun handleBadRequest(ex: BadRequestException): ResponseEntity<ErrorResponse> {
        return ResponseEntity(ErrorResponse(ex.message ?: "Bad Request", ERR_BAD_REQUEST), HttpStatus.BAD_REQUEST)
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidation(ex: MethodArgumentNotValidException): ResponseEntity<ErrorResponse> {
        val message = ex.bindingResult.allErrors.joinToString("; ") { it.defaultMessage ?: "Error" }
        return ResponseEntity(ErrorResponse(message, ERR_VALIDATION), HttpStatus.BAD_REQUEST)
    }

    @ExceptionHandler(MethodArgumentTypeMismatchException::class)
    fun handleTypeMismatch(ex: MethodArgumentTypeMismatchException): ResponseEntity<ErrorResponse> {
        return ResponseEntity(ErrorResponse(ex.message ?: "Type mismatch", ERR_BAD_REQUEST), HttpStatus.BAD_REQUEST)
    }

    @ExceptionHandler(HttpMessageNotReadableException::class)
    fun handleMessageNotReadable(ex: HttpMessageNotReadableException): ResponseEntity<ErrorResponse> {
        return ResponseEntity(ErrorResponse("Message not readable (bad formatting)", ERR_BAD_REQUEST), HttpStatus.BAD_REQUEST)
    }

    @ExceptionHandler(RestClientResponseException::class)
    fun handleRemote(ex: RestClientResponseException): ResponseEntity<String> {
        return ResponseEntity.status(ex.statusCode).body(ex.responseBodyAsString)
    }


    @ExceptionHandler(Exception::class)
    fun handleGeneral(ex: Exception): ResponseEntity<ErrorResponse> {
        return ResponseEntity(ErrorResponse(ex.message ?: "Internal Error", ERR_INTERNAL), HttpStatus.INTERNAL_SERVER_ERROR)
    }

    companion object {
        const val ERR_NOT_FOUND = 40401
        const val ERR_FORBIDDEN = 40301
        const val ERR_BAD_REQUEST = 40001
        const val ERR_VALIDATION = 40002
        const val ERR_INTERNAL = 50001

    }
}
